### PR TITLE
Convert binary decimal

### PR DESCRIPTION
**Pull request by Tahershaba**
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations
- [ ] The function is not called in the function file

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by sara-shahin-cell** [comment]
Hi Taher
To fix CI checks you need to 
# 1- install ruff
pip install ruff
# 2- run ruff format on terminal
ruff format.
# 3- run a ruff check to check if everything is ok
ruff check

File: solutions/Convert_Binary_To_Decimal.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+"""Summery
```
Comment (sara-shahin-cell) [inline solutions/Convert_Binary_To_Decimal.py:1]: there is no module docstring in the function file. pls add this in the first line :
```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
```

**Comment by Fikre-Mamo** [comment]
# Code Review for `conv_bin_dec`

### **What’s Great About Your Code**
1. **Clear Functionality**:  
   The `conv_bin_dec` function works as expected, converting binary strings to decimal integers. It efficiently handles spaces and invalid characters through assertions.

2. **Unit Tests**:  
   Your unit tests are well-structured, testing the function against valid binary inputs, invalid inputs, and edge cases like spaces in the binary string.

3. **Error Handling**:  
   You have included a check for non-binary characters using `assert`, which is a good way to prevent invalid inputs. It’s a nice touch that the assertion raises a helpful error message.

---

### **Suggestions to Polish and Improve Your Code**

#### **Behavior and Structure**
1. **File Names**:  
   Ensure that your Python file names are lowercase with underscores, such as `convert_binary_to_decimal.py` to follow the naming convention. Similarly, your test file can be named `test_convert_binary_to_decimal.py`.

2. **Module Docstring in Test File**:  
   Adding a brief docstring at the top of the test file will help maintain clarity. For example:
   ```python
   """
   Tests for the conv_bin_dec function to ensure correct binary to decimal conversion.
   """

### ruff format
Insure to install and run ruff become its help you to write clean and standard python code

File: solutions/Convert_Binary_To_Decimal.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    >>> conv_bin_dec("00000110")
+    6
+    >>> conv_bin_dec("00001111")
+    15
+    >>> conv_bin_dec("00010010")
+    18
+    >>> conv_bin_dec("001r001")
+    AssertionError:"Argument is not a binary number."
+    """
```
Comment (sara-shahin-cell) [inline solutions/Convert_Binary_To_Decimal.py:29]: There are no defensive assertions in the code. please add it.
ex:
` assert isinstance (bin, (str)), "Input must be a string (str)." `

**Review by sara-shahin-cell — CHANGES_REQUESTED** [review]
# Behavior

## Files

- [x] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [ ] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

## Strategy

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

File: solutions/tests/Test_convert_binary_to_decimal.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+"""
```
Comment (KimaciaJnr) [inline solutions/tests/Test_convert_binary_to_decimal.py:1]: Like the solution file, please ensure the test file has a module docstring too.
Example;
                      #!/usr/bin/env python3
                      # -*- coding: utf-8 -*-
